fix compilation on an x32 (amd64ilp32) host system again

mirabilos opened the pull request:
- fix libressl build for x32 by disabling assembly code there; x32 is mis-detected as amd64, so it uses wrong asm code
- unbreak content of the GNU_HOST_NAME variable that got broken during some refactoring; `config.guess` also mis-detects x32 as amd64 (because the maintainers deliberately do _not_ want to call the host compiler to check whether it’s x32 or not)

mirabilos committed:
build: fix GNU_HOST_NAME detection

unbreaks compilation on targets such as x32 (amd64ilp32)
which are mis-detected by config.guess (unfortunately,
even the last update doesn’t fix this because it tries
to guess by the format of /bin/sh instead of asking the
host compiler)

Fixes: 9467ce42da7f (“build: get rid of host.mk”)

Signed-off-by: mirabilos <m@mirbsd.org>

mirabilos commented:
I also ran the libressl testsuite successfully, I should note.

mirabilos commented:
I added a third commit to disable even attempting to build the qemu host tool; it’s so ancient that getting it to work on x32 isn’t worth the effort, and the user can just install their distro’s qemu anyway.

hauke commented on include/toplevel.mk:
```diff
@@ -25,7 +25,7 @@ export SOURCE_DATE_EPOCH
 export GIT_ASKPASS:=/bin/true
 export MAKE_JOBSERVER=$(filter --jobserver%,$(MAKEFLAGS))
-export GNU_HOST_NAME:=$(shell $(TOPDIR)/scripts/config.guess)
+export GNU_HOST_NAME:=$(shell $(TOPDIR)/scripts/get-gnu-host-name.sh)
```
I think we just better try to fix config.guess. It looks like this commit it missing:
https://git.savannah.gnu.org/gitweb/?p=config.git;a=commitdiff;h=bd9626458c30d7faec17d7dfbd85a80617b10007

I would prefer if we just update config.guess and config.sub to the most recent version

hauke commented:
I merged your first patch and updated the config.guess and config.sub 
Please rebase your patches and try again how much is working now.

mirabilos commented:
Hi Hauke,

>I think we just better try to fix config.guess. It looks like this commit it missing:
>https://git.savannah.gnu.org/gitweb/?p=config.git;a=commitdiff;h=bd9626458c30d7faec17d7dfbd85a80617b10007

this commit will not work, for reasons I outlined in the thread around
https://lists.nongnu.org/archive/html/config-patches/2018-02/msg00019.html
(mostly: x32 is normally used in a Multi-Arch scenario; additionally,
one of the /bin/sh candidates is built using klibc, whose x32 binaries
detect as amd64, not x32).

This is the patch you’d want:

https://lists.nongnu.org/archive/html/config-patches/2018-02/msg00014.html

Unfortunately, the GNU people decided to ban CC_FOR_BUILD, preventing
us from using the only proper method to detect x32 hosts.

mirabilos commented:
Hauke Mehrtens dixit:

>Please rebase your patches and try again how much is working now.

It won’t, as I wrote in the other mail. Proof:

tglase@tglase:~/stuff/openwrt $ ./scripts/config.guess
x86_64-pc-linux-gnu

(after a “git pull”)

adrianschmutzler commented:
@mirabilos Looks like the last status is "not working". Do you plan on updating anything here? Otherwise, we will close this.

mirabilos commented:
@adrianschmutzler what do you mean “not working”?

Last time I updated it, it worked very well.

adrianschmutzler commented:
>> Please rebase your patches and try again how much is working now.
> It won’t, as I wrote in the other mail. Proof:

This sounds to me like you saying your rebased patches are not working.

Anyway, since last post is from 2018, you might want to explain your latest changes so hauke can respond if he's interested.

adrianschmutzler commented:
@hauke Maybe have another look so we can sort out this old PR.
@mirabilos Some minor formal issues:
In first commit, you should use "Fixes: " instead of "broken in commit" before the commit number.
For second commit, please make the commit title more precise (what is "a qemu this ancient") and be more chatty about what you do and why.
I do not think there are issues with your English, it was just the brevity of your answer that made it easy to be misunderstood.

mirabilos commented:
@adrianschmutzler ok, I’ve changed the commit message and tried to explain the qemu one better (0.14 is positively ancient, and x32 support came in much, much later, so we just skip it, and when a user on x32 wants qemu they just to have install one from their distro), and since I was there anyway, rebased on current master.

adrianschmutzler commented:
Thanks. Just had another look: Your From: doesn't match your Signed-off-by:
This can be fixed for each commit by "git commit --amend --reset-author".
If you change anyway, the "Fixes" can be improved by adding the commit title after the 12-digit commit id:
Fixes: 9467ce42da7f ("build: get rid of host.mk")

mirabilos commented:
Ah yes, that Signed-off-by was from before I harmonised my git configuration. Fixed, and the commit title also added.

adrianschmutzler commented:
Signed-off-by has to be a real name, and thus the same applies to From ...

mirabilos commented:
I consider “mirabilos” my real name, and am working at getting it added to my ID card.

In the UK, people can change their real name to something like that, e.g. there’s a Debian Developer called “Wookey”, which is their legal name.

mirabilos commented:
Rebased on latest master, since I’m looking at the MIPS16 issue.

aparcar commented:
Please rebase. I'll then try to run that in a 32bit VM

mirabilos commented:
Paul Spooren dixit:

>Please rebase. I'll then try to run that in a 32bit VM

x32 is a 64-bit system. You cannot run x32 binaries in a 32-bit VM.

mirabilos commented:
I’m compiling this now (good chance to update the packages I contributed again as well) to test.

mirabilos commented:
It builds and boots.

CodeFetch commented:
@mirabilos Please rebase.
@hauke It does not look like this could break anything. What's the problem here?

aparcar commented on tools/Makefile:
```diff
@@ -33,7 +33,9 @@ tools-$(CONFIG_TARGET_apm821xx)$(CONFIG_TARGET_gemini) += genext2fs
 tools-$(CONFIG_TARGET_mxs) += elftosb sdimage
 tools-$(CONFIG_TARGET_tegra) += cbootimage cbootimage-configs
+ifneq ($(GNU_HOST_NAME),x86_64-linux-gnux32)
 tools-$(CONFIG_TARGET_x86) += qemu
```
Qemu is no longer maintained within openwrt.git, this can be removed.

aparcar commented:
Would another update solve the issue? Please test https://github.com/openwrt/openwrt/pull/4207

mirabilos commented:
Paul Spooren dixit:

>Would another update solve the issue? Please test
>https://github.com/openwrt/openwrt/pull/4207

I know it won’t. I’ve been talking with the developers of
config.{guess,sub}, and they positively deprecated using
the compiler to test for the architecture, which, in the
case of x32 (and similar “64-bit ILP32” architectures like
the upcoming arm64ilp32 and possibly more), is necessary
because they’re “just” a different userspace ABI, not a
kernel architecture. So basically the compiler is the sole
source of the truth we are seeking. The GNU config tools
ask file(1) to determine the bit-ness of a compiled utility
instead, which is fragile especially as klibc/x32 actually
is 64-bit so the test regularily fails :/

This means the part changing GNU_HOST_NAME is really needed.

The in-tree qemu must be excluded on x32 because it is so
old it predates x32 support. If you’d rather entirely remove
it, fine with me, please do so then I’ll rebase this as
Vincent asked.

bye,
//mirabilos
-- 
„Cool, /usr/share/doc/mksh/examples/uhr.gz ist ja ein Grund,
mksh auf jedem System zu installieren.“
	-- XTaran auf der OpenRheinRuhr, ganz begeistert
(EN: “[…]uhr.gz is a reason to install mksh on every system.”)

mirabilos commented:
@CodeFetch rebased

CodeFetch reviewed (approved):
Looks good to me.

@hauke @aparcar Is it okay?

KanjiMonster commented:
Looking at config.guess upstream changes, https://git.savannah.gnu.org/gitweb/?p=config.git;a=commitdiff;h=f56a7140386d08a531bcfd444d632b28c61a6329 seems to tackle the issue, and  https://git.savannah.gnu.org/gitweb/?p=config.git;a=commitdiff;h=af8d803a82436779d35ea389888788c78677804e;hp=602766470c886df7ae07bcfd7dcf532f0783d3e0 for arm64.

So an update to current HEAD might be enough and avoid this extra check. Can you (still) test?

mirabilos commented:
Jonas Gorski dixit:

>Looking at config.guess upstream changes,
>https://git.savannah.gnu.org/gitweb/?p=config.git;a=commitdiff;h=f56a7140386d08a531bcfd444d632b28c61a6329
>seems to tackle the issue, and
>https://git.savannah.gnu.org/gitweb/?p=config.git;a=commitdiff;h=af8d803a82436779d35ea389888788c78677804e;hp=602766470c886df7ae07bcfd7dcf532f0783d3e0
>for arm64.

Oh, wow. They went back to using CC_FOR_BUILD (which they had
refused before).

This patch looks good and will correctly catch x32. I’m not
informed about the various triplets for ARM, and arm64ilp32
will be a thing, but it seems that upstream can DTRT now. I
had reported this here because they initially refused that.

Unfortunately I cannot easily test any more, but visual
inspection shows it’s sufficient.

bye,
//mirabilos
-- 
Solange man keine schmutzigen Tricks macht, und ich meine *wirklich*
schmutzige Tricks, wie bei einer doppelt verketteten Liste beide
Pointer XORen und in nur einem Word speichern, funktioniert Boehm ganz
hervorragend.		-- Andreas Bogk über boehm-gc in d.a.s.r

neheb commented:
Is this not needed anymore?

mirabilos commented:
Since `scripts/config.guess` is [now recent enough](https://github.com/openwrt/openwrt/blob/4145ff4d8a29c1c7a1569bb06fa4d1fe9808c94f/scripts/config.guess#L1118), no (at least not in the master branch).
